src/plugins/backend_odbc/backend_odbc_get.c

```diff
@@ -0,0 +1,827 @@
+
+
+#include "./backend_odbc_get.h"
+
+#include <kdbassert.h>
+#include <kdberrors.h>
+#include <kdbhelper.h>
+#include <kdblogger.h>
+#include <string.h>
+
+
+/**
+ * @brief Constructs a SELECT query string for the ODBC backend based on the given data source configuration
+ *
+ * @param dsConfig A valid data source config, as returned by the \ref fillDsStructFromDefinitionKs "fillDsStructFromDefinitionKs" function
+ * @param quoteString The characters that should be added before and after identifiers, pass NULL if your identifiers in @dsconfig are
+ * 	already quoted or if you don't want to use quoted identifiers
+ *
+ * @return The query string for the select query to get keynames, key-values and metadata from an SQL data source.
+ * 	Make sure to free the returned string.
+ * @retval "" (empty string) if invalid input detected
+ * 	This string must not be freed!
+ * @retval NULL if memory allocation failed
+ */
+static char * getSelectQueryString (struct dataSourceConfig * dsConfig, char * quoteString)
+{
+	/* A sample query string that shows the structure of the SELECT query that this function generates:
+	 * SELECT "elektra"."key", "elektra"."val", "elektrameta"."metakey", "elektrameta"."metaval" FROM {oj "elektra" LEFT OUTER JOIN
+	 * "elektrameta" ON "elektra"."key"="elektrameta"."key"}
+	 */
+
+	/* Verify that all necessary strings are provided */
+	if (!dsConfig || !(dsConfig->tableName) || !(*(dsConfig->tableName)) || !(dsConfig->keyColName) || !(*(dsConfig->keyColName)) ||
+	    !(dsConfig->valColName) || !(*(dsConfig->valColName)) || !(dsConfig->metaTableName) || !(*(dsConfig->metaTableName)) ||
+	    !(dsConfig->metaTableKeyColName) || !(*(dsConfig->metaTableKeyColName)) || !(dsConfig->metaTableMetaKeyColName) ||
+	    !(*(dsConfig->metaTableMetaKeyColName)) || !(dsConfig->metaTableMetaValColName) || !(*(dsConfig->metaTableMetaValColName)))
+	{
+		return "";
+	}
+
+
+	/* 1. Calculate strlen of all column names */
+	size_t sumLen = strlen (dsConfig->keyColName) * 2 + strlen (dsConfig->valColName) + strlen (dsConfig->metaTableMetaKeyColName) +
+			strlen (dsConfig->metaTableMetaValColName) + strlen (dsConfig->metaTableKeyColName);
+
+	/* 2. Add table names (for SELECT and outer join parts of the query string, add 6 bytes for the cases where the column name follows
+	 * the table name ('.' as separator */
+	sumLen += strlen (dsConfig->tableName) * 4 + strlen (dsConfig->metaTableName) * 4 + 6;
+
+	/* 3. Add strlen for static parts of the query + 1 byte for \0 */
+	sumLen += strlen ("SELECT , , ,  FROM {oj  LEFT OUTER JOIN  ON =}");
+
+	/* 4. Add bytes for quoting identifiers (table- and column-names) */
+	if (quoteString)
+	{
+		sumLen += (14 * 2 * strlen (quoteString));
+	}
+	else
+	{
+		/* Don't use quotes for identifiers */
+		quoteString = "";
+	}
+
+	char * queryString = elektraMalloc (sizeof (char) * sumLen);
+
+	if (!queryString)
+	{
+		return NULL;
+	}
+
+	/* Build the string */
+	char * strEnd = stpcpy (queryString, "SELECT ");
```

Comment:
I thought so too, but at least the ODBC-API doesn't allow parameters for the column list of a SELECT-query. :(
In the [ODBC documentation](https://learn.microsoft.com/en-us/sql/odbc/reference/develop-app/statement-parameters) is written:
> Parameters are valid only in certain places in SQL statements. For example, they are **not allowed in the select list (the list of columns to be returned by a SELECT statement)**, nor are they allowed as both operands of a binary operator such as the equal sign (=), because it would be impossible to determine the parameter type.

So the escaping of the identifiers has to happen somewhere else and independent of the used ODBC driver (and therefore data source like different DBMS or also file-based formats).

One option would be to make quoted identifiers obligatory and reject parameters that already contain quote-characters.

Reply:
I've added the check for detecting `SQL_IDENTIFIER_QUOTE_CHAR` in the identifiers in `dsConfig` and made using `SQL_IDENTIFIER_QUOTE_CHAR` obligatory.

The concept with the SQL-queries instead of table- and column names is an idea for a follow-up PR, but not in the scope of this PR now.